### PR TITLE
Fixes #9480 - multiple NICs integration with compute profiles

### PR DESCRIPTION
**Pull request by tstrachota**
From users point of view:
- interfaces setup in compute profiles is back
- interface related compute attrs get merged into host NICs upon compute
  profile selection
- NIC overview table displays details in the column "Type"
- UI for direct creation of VMs fixed
- fixed removing interfaces and volumes from compute profiles

From technical point of view:
- NIC overview table is always built by js
- templates for compute resources got more structure (split to partials
  `_base`, `_interface`, `_volume`)
- compute resources can provide custom implementation of
  `providerSpecificNICInfo` function

### CODE REVIEW

**Comment by tstrachota** [comment]
@ares @GregSutcliffe please take a look.

**Comment by domcleal** [comment]
Administrative, can you see if #9222 in Redmine is now a dupe and if this fixes any of the other issues on the 1.8 bug list I sent to foreman-dev?  (Maybe #9325?)

File: app/views/hosts/_compute_detail.html.erb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+<%= render :partial => provider_partial(compute_resource, 'base'),
+            :locals => { :f => f, :compute_resource => compute_resource, :new_host => new_host }.merge(args_for_compute_resource_partial(@host)) %>
+
+<!--NICS-->
+<% if provider_partial_exist?(compute_resource, 'network') %>
+  <fieldset>
+    <legend><%= _("Network interfaces") %></legend>
+    <p class="help-block">
+      <%= _("Network interfaces management has been moved to the Interfaces tab.") %>
+      <%= _("Please set your interfaces") %> <a href="#" onclick="$('#network_tab a').click(); return false;"><%= _("there") %></a>.
```
Comment (domcleal) [inline app/views/hosts/_compute_detail.html.erb]: i18n - this is going to be very hard for anybody to translate (I realise it was already merged, but it wasn't picked up then).  One reason is that the words are separate, there's no link in the string extraction that it's continuation of a phrase, but also because the space between the words has to be part of the string for eastern languages.

I'd either come up with an entire phrase that you can make a link, remove the whole thing, or maaaybe use %{link_start} and %{link_end} to insert the pieces of HTML into the translated string (and some .html_safe).

**Comment by tstrachota** [comment]
@domcleal correct #9222 was a dupe. #9325 has been fixed separately.

**Comment by ares** [comment]
@tstrachota see failing tests http://ci.theforeman.org/job/test_develop_pr_core/4124/

File: app/assets/javascripts/host_edit_interfaces.js [hunk 2]
```diff
@@ -253,6 +265,43 @@ $(document).on('click', '.provision-flag', function () {
...
+    if (typeof(providerSpecificNICInfo) == "function")
+      info = providerSpecificNICInfo(form);
+    else
+      info = __("physical")
+
+  }
+  return info;
+}
+
+// TODO: add on text field change
```
Comment (ares) [inline app/assets/javascripts/host_edit_interfaces.js]: is that still valid TODO? if so, could you elaborate what it means?

File: app/views/compute_resources_vms/form/libvirt/_base.html.erb [hunk 3]
```diff
@@ -36,3 +21,13 @@
                :help_block  => _("Image to use"),
                :label => _('Image')} %>
 </div>
+
+<script>
```
Comment (ares) [inline app/views/compute_resources_vms/form/libvirt/_base.html.erb]: could this be extracted into app/assets and included using `<%= javascript_include_tag '...libvirt...' %>`

EDIT: or at least the function if you want to keep providerSpecificNICInfo for the context

File: app/services/interface_merge.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+class InterfaceMerge
```
Comment (ares) [inline app/services/interface_merge.rb:1]: :+1: for extracting into class

File: app/services/interface_merge.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  def run(interfaces, compute_attrs)
+    vm_interfaces = compute_attrs.vm_interfaces
+
+    # merge with existing
+    interfaces.each do |nic|
+      if !nic.virtual?
```
Comment (ares) [inline app/services/interface_merge.rb]: maybe this would be more readable

``` ruby
interfaces.select(&:physical?).each do |nic|
...
end
```

which requires adding

``` ruby
def physical?
  !virtual?
end
```

to interface, this might be handy elsewhere

File: app/services/interface_merge.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        vm_nic = vm_interfaces.shift
+        return if vm_nic.nil?
+        nic.compute_attributes = vm_nic
```
Comment (ares) [inline app/services/interface_merge.rb]:
``` ruby
nic.compute_attributes = vm_nic
nic.compute_attributes["from_profile"] = compute_attrs.compute_profile.name
```

these lines could be extracted to some private `sync(nic, vm_nic, profile_name)` method so you don't repeat it later when creating new interfaces below

compute_attrs.compute_profile.name could be extracted to profile_name var

also use single quoted strings if there's no interpolation

File: test/unit/interface_merge_test.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    {
+      'attr' => num,
+      'from_profile' => @profile_name
+    }
+  end
+
+  test "it merges compute attributes with existing NICs" do
+    interfaces = [
+      Nic::Managed.new(:identifier => 'eth0'),
+      Nic::Managed.new(:identifier => 'eth1'),
```
Comment (ares) [inline test/unit/interface_merge_test.rb]: why not using FactoryGirl? (in all tests where applicable)

File: test/unit/interface_merge_test.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    ]
+    @merge.run(interfaces, @attributes)
+
+    interfaces.length.must_equal 2
+    interfaces[0].compute_attributes.must_equal(attrs(1))
+    interfaces[0].identifier.must_equal('eth0')
+
+    interfaces[1].compute_attributes.must_equal(attrs(2))
+    interfaces[1].identifier.must_equal(nil)
+  end
```
Comment (ares) [inline test/unit/interface_merge_test.rb:70]: tests do not cover setting profile name, otherwise I think it covers it well

File: app/controllers/hosts_controller.rb [hunk 9]
```diff
@@ -135,6 +135,15 @@ def compute_resource_selected
 
   def interfaces
     @host = Host.new params[:host]
+
+    if @host.compute_profile_id && @host.compute_resource_id
+      profile = @host.compute_resource.compute_attributes.find_by_compute_profile_id(@host.compute_profile_id)
+      if profile
+        merge = InterfaceMerge.new
```
Comment (ares) [inline app/controllers/hosts_controller.rb]: just and idea... would it make more sense to pass just @host so InterfaceMerge would do the merging and returned host with merged interfaces? this `@host.interfaces` and compute_profile logic should not live in controller I think

File: app/models/compute_attribute.rb [hunk 10]
```diff
@@ -30,4 +34,13 @@ def pretty_vm_attrs
   def update_name
     self.name = pretty_vm_attrs if pretty_vm_attrs.present?
   end
+
+  def attribute_values(attr_name)
+    attr_key = "#{attr_name}_attributes"
+    if vm_attrs[attr_key]
+      vm_attrs[attr_key].select{|k,v| k.to_s != "new_#{attr_name}" }.values
```
Comment (ares) [inline app/models/compute_attribute.rb]: nit, spaces around `{`

File: app/views/compute_resources_vms/form/_networks.html.erb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        :layout => 'compute_resources_vms/form/deletable_layout' }) %>
+
+    <%= field_set_tag _("Network interfaces"), :id => "network_interfaces" do %>
+
+      <% if compute_resource.networks.any? %>
+        <%= f.fields_for compute_resource.interfaces_attrs_name do |i| %>
+
+          <%= render :partial => provider_partial(compute_resource, 'network'),
+          :locals => { :f => i, :compute_resource => compute_resource, :new_host => new_host, :remove_title => _('remove network interface') },
+          :layout => 'compute_resources_vms/form/deletable_layout' %>
```
Comment (ares) [inline app/views/compute_resources_vms/form/_networks.html.erb]: nit, indentation looks broken

File: app/views/compute_resources_vms/form/libvirt/_network.html.erb [hunk 12]
```diff
@@ -1,36 +1,30 @@
...
-  <%= f.fields_for 'compute_attributes', OpenStruct.new(f.object.compute_attributes) do |f| %>
+<% nat    = compute_resource.networks %>
+<% bridge = compute_resource.interfaces %>
 
-    <%=
-    selectable_f f, :type, libvirt_networks(compute_resource), {},
-                     { :label    => _('Network type'),
-                       :disabled => disabled,
-                       :onchange => 'libvirt_network_selected(this)' } %>
+<%=
```
Comment (ares) [inline app/views/compute_resources_vms/form/libvirt/_network.html.erb]: I see it's not your code but could you put `<%=` on same line with `selectable_f` please?

File: app/views/usergroups/_form.html.erb [hunk 13]
```diff
@@ -54,7 +54,7 @@
         </div>
         <div class="children_fields">
           <%= new_child_fields_template(f, :external_usergroups, { :partial => 'external' } ) %>
-          <%= add_child_link '+ ' + _('Add external user group'), :external_usergroups, { :class => "info disable-unsupported", :title => _('link external user group with this user group') } %>
+          <%= add_child_link '+ ' + _('Add external user group'), :external_usergroups, { :class => "info", :title => _('link external user group with this user group') } %>
```
Comment (ares) [inline app/views/usergroups/_form.html.erb:58]: how is this related? I see we use this class in host_edit.js, but this is user group? is that unrelated cleanup?

**Comment by ares** [comment]
When I go into compute profiles, select some libvirt compute resource getting 

```
NameError in Compute_attributes#edit
undefined local variable or method `disabled' for #<#<Class:0x007f77a8857ec0>:0x007f77a8855940>
in app/views/compute_resources_vms/form/libvirt/_network.html.erb where line #23
```

**Comment by ares** [comment]
In interfaces modal window, when I click virtual NIC, computer attributes are hidden (expected) but no virtual fields (tag, attached_to) are displayed. No error in JS console.

EDIT: seems to happen only for new interface, when I change type to BOND and back it starts to work as expected

File: test/unit/interface_merge_test.rb (hunk 8 above)
Comment (tstrachota) [inline test/unit/interface_merge_test.rb:70]: Setting profile name is covered by comparing with `attrs`. See line 10 of the test. I'll rename the method to `expected_attrs` to make it more obvious.

File: app/views/usergroups/_form.html.erb (hunk 13 above)
Comment (tstrachota) [inline app/views/usergroups/_form.html.erb:58]: The class obviously got here just by copy-pasting. We used `disable-unsupported` only in hosts but it's no longer necessary. I replaced it with variable `new_host?` in the partials.
I see I forgot to remove the jQuery piece. I'll do it with other updates of the PR.

**Comment by tstrachota** [comment]
@ares updated according to your comments. Still TODO:
- failing test
- consider changing `InterfaceMerge` to use host object

File: test/unit/nic_test.rb [hunk 14]
```diff
@@ -107,6 +107,18 @@ def teardown
     refute_includes nic.errors.keys, :identifier
   end
 
+  context 'physial?' do
```
Comment (ares) [inline test/unit/nic_test.rb]: typo `physial?` :-)

File: test/unit/interface_merge_test.rb [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+  test "it merges compute attributes with existing NICs" do
+    interfaces = [
+      FactoryGirl.build(:nic_managed, :identifier => 'eth0'),
+      FactoryGirl.build(:nic_managed, :identifier => 'eth1'),
+      FactoryGirl.build(:nic_managed, :identifier => 'eth2')
+    ]
+    @merge.run(interfaces, @attributes)
+
+    interfaces.length.must_equal 3
```
Comment (ares) [inline test/unit/interface_merge_test.rb]: while I like this syntax more, we use assert_equal on whole foreman unit tests

**Comment by tstrachota** [comment]
@ares updated according to your comments.
I considered `InterfaceMerge` using host but decided to keep the input attributes simple at the end. I moved the nil check into the `run` though. The controller method is simpler now.

**Comment by ares** [comment]
@tstrachota starting to test this, could you rebase meanwhile?

**Comment by ares** [comment]
One issue while testing with libvirt. When I add a physical NIC, set it as bridge to em1, I see correct note in type column `physical @ bridge em1` but when I reedit the NIC, it's Network type select box shows it's NAT, while Network is set to em1 (so the form is mixed).

File: test/unit/compute_attribute_test.rb [hunk 16]
```diff
@@ -33,4 +33,19 @@ class ComputeAttributeTest < ActiveSupport::TestCase
...
+      set = compute_attributes(:with_interfaces)
+      set.vm_interfaces.must_equal [
+        {'attr' => 1},
+        {'attr' => 2}
+      ]
+    end
+
+    test "returns empty array if interface attributes are missing" do
+      set = compute_attributes(:one)
+      set.vm_interfaces.must_equal []
```
Comment (ares) [inline test/unit/compute_attribute_test.rb]: assert_equal?
EDIT: or assert_empty

File: test/unit/compute_attribute_test.rb [hunk 17]
```diff
@@ -33,4 +33,19 @@ class ComputeAttributeTest < ActiveSupport::TestCase
       @set.nonexistant_flavor_field
     end
   end
+
+  describe "vm_interfaces" do
+    test "returns array of interface attributes" do
+      set = compute_attributes(:with_interfaces)
+      set.vm_interfaces.must_equal [
```
Comment (ares) [inline test/unit/compute_attribute_test.rb]: assert_equal?

File: test/unit/nic_test.rb [hunk 18]
```diff
@@ -107,6 +107,18 @@ def teardown
     refute_includes nic.errors.keys, :identifier
   end
 
+  context 'physical?' do
+    test 'returns true for a physical interface' do
+      nic = FactoryGirl.build(:nic_managed, :virtual => false)
+      nic.physical?.must_equal true
```
Comment (ares) [inline test/unit/nic_test.rb]: assert_equal?

File: test/unit/nic_test.rb [hunk 19]
```diff
@@ -107,6 +107,18 @@ def teardown
...
+    end
+
+    test 'returns false for a virtual interface' do
+      nic = FactoryGirl.build(:nic_managed, :virtual => true)
+      nic.physical?.must_equal false
```
Comment (ares) [inline test/unit/nic_test.rb]: assert_equal?

File: test/unit/nic_test.rb (hunk 18 above)
Comment (ares) [inline test/unit/nic_test.rb]: sorry, just `assert nic.physical?`

File: test/unit/nic_test.rb (hunk 19 above)
Comment (ares) [inline test/unit/nic_test.rb]: sorry, just `refute nic.physical?`

**Comment by ares** [comment]
@tstrachota other than few nits in new code and an issue I found during testing it works pretty amazing!

**Comment by tstrachota** [comment]
@ares rebased. Sorry for the overlooked must_equals. The select box issue is fixed too.

**Comment by GregSutcliffe** [comment]
Just done some light testing. UI is awesome, @tstrachota++ for that. Not sure about hammer though. I used:

```
bundle exec bin/hammer host create \                                                 
  --name "foo" \                                                           
  --hostgroup-id 2 \                                                                 
  --operatingsystem-id 1 \                                                         
  --subnet-id 1 \                                                                    
  --ip "192.168.130.7" \                                                                     
  --compute-resource-id 1 \                                                          
  --compute-profile-id 1 \                                                           
  --provision-method image                                                           
```

But the network was wrongly created: 

```
#<Nic::Managed id: nil, mac: nil, ip: "192.168.130.7", ..., compute_attributes: {}> 
```

I suspect this is probably a NIC API issue rather than a problem with the compute profiles though.

**Comment by tstrachota** [comment]
@GregSutcliffe api is most likely broken since the first multi NIC PR. I'm playing with it right now. We should fix it in a separate PR.

File: app/models/nic/base.rb [hunk 20]
```diff
@@ -73,6 +73,10 @@ class Jail < ::Safemode::Jail
             :attached_devices_identifiers, :primary, :provision, :inheriting_mac
     end
 
+    def physical?
```
Comment (ares) [inline app/models/nic/base.rb:76]: sorry I forgot to mention this previously, could you add this to Jail above please?

**Comment by GregSutcliffe** [comment]
:+1: from testing. I've not reviewed though (@ares seems to have that covered :P)

**Comment by ares** [comment]
Merged as https://github.com/theforeman/foreman/commit/6d05514ad02676067389f44c4ba1d3daf357884b, thanks @tstrachota!
